### PR TITLE
(feat): Support for groups

### PR DESCRIPTION
**Pull request by aramirezmartin**

### CODE REVIEW

File: vt_graph_api/graph.py [hunk 1]
```diff
@@ -1750,12 +1869,88 @@ def download_screenshot(self, path = "."):
...
+        group_links.append(new_link)
+    return group_links
+
+
+  def _get_special_relationship_nodes(self):
+    """Returns a list with the special relationship nodes."""
+    return list(self.special_relationship_nodes.values())
+
+
+  def _get_special_relationship_links(self, final_nodes):
```
Comment (dprolongo) [inline vt_graph_api/graph.py:1940]: _get_special_relationship_nodes retrieves nodes from self. In the case of links we are receiving a parameter. Are they from the same flow?

File: vt_graph_api/graph.py [hunk 2]
```diff
@@ -1805,17 +2000,68 @@ def create_collection(self, name=None, description=None):
...
     """
-    self.representation = representation
 
+    Args:
+      node_ids:
+      group_name:
 
+    Returns:
 
+    Raises: CreateGroupError if the
```
Comment (dprolongo) [inline vt_graph_api/graph.py:2017]: Incomplete, I guess

File: vt_graph_api/graph.py [hunk 3]
```diff
@@ -1805,17 +2000,68 @@ def create_collection(self, name=None, description=None):
...
+            "Node {node_id} is not in the Graph.".format(
+                node_id=node_id))
+
+    # Check if nodes are already in a group.
+    nodes_already_in_a_group = set()
+    for group_node in self.group_nodes.values():
+      for node_id in group_node["entity_attributes"]["grouped_node_ids"]:
+        nodes_already_in_a_group.add(node_id)
+
+    intersection = nodes_already_in_a_group.intersection(node_ids)
```
Comment (dprolongo) [inline vt_graph_api/graph.py:2041]: given the cardinality if the nodes to add to a group will be usually lower than nodes in current groups, what about checking if the nodes in other groups are already in node_ids (may be a set) and raise the error as soon as possible, at line 2039?

File: vt_graph_api/graph.py [hunk 4]
```diff
@@ -1750,12 +1869,88 @@ def download_screenshot(self, path = "."):
...
         {'type': node.node_type, 'id': node.node_id}
         for node in self.nodes.values() if node.node_type == node_type
     ]}
 
+
+  def _get_groups_nodes(self, final_nodes):
+    """Returns all the loaded and generated group nodes.
+
+    Args:
+      final_nodes: Set with the ids of all the added nodes.
```
[the remark this answers is not in the document]
Comment (aramirezmartin) [reply vt_graph_api/graph.py:1884]: Yes, it is supposed to get the group nodes once the graph is going to be saved.

File: vt_graph_api/graph.py [hunk 5]
```diff
@@ -1750,12 +1869,88 @@ def download_screenshot(self, path = "."):
...
+    """
+    group_nodes = []
+    current_index = len(final_nodes)
+    for idx, group_node in enumerate(self.group_nodes.values()):
+      belonging_nodes = group_node['entity_attributes']['grouped_node_ids']
+      belonging_nodes = list(
+          set([node for node in belonging_nodes if node in final_nodes]))
+      if (belonging_nodes):
+        group_node['entity_attributes']['grouped_node_ids'] = belonging_nodes
+        group_node['index'] = current_index + idx
```
[the remark this answers is not in the document]
Comment (aramirezmartin) [reply vt_graph_api/graph.py:1897]: Index is used in VT internally to get the order in which the node is inserted.

**Review by dprolongo — APPROVED** [review]

**Pull request merged**